contracts/margin/external/exchangewrappers/ZeroExV2ExchangeWrapper.sol

```diff
@@ -0,0 +1,295 @@
+import { TokenInteract } from "../../../lib/TokenInteract.sol";
+import { ExchangeReader } from "../../interfaces/ExchangeReader.sol";
+import { ExchangeWrapper } from "../../interfaces/ExchangeWrapper.sol";
+
+
+/**
+ * @title ZeroExV2ExchangeWrapper
+ * @author dYdX
+ *
+ * dYdX ExchangeWrapper to interface with 0x Version 2
+ */
+contract ZeroExV2ExchangeWrapper is
+    LibFillResults,
+    LibOrder,
+    ExchangeWrapper,
+    ExchangeReader
+{
+    using SafeMath for uint256;
+    using TokenInteract for address;
+
+    // ============ State Variables ============
+
+    // msg.senders that will put the correct tradeOriginator in callerData when doing an exchange
+    mapping (address => bool) public TRUSTED_MSG_SENDER;
+
+    // address of the ZeroEx V1 Exchange
+    address public ZERO_EX_EXCHANGE;
+
+    // address of the ZeroEx V1 TokenTransferProxy
+    address public ZERO_EX_TOKEN_PROXY;
+
+    // address of the ZRX token
+    address public ZRX;
+
+    // ============ Constructor ============
+
+    constructor(
+        address zeroExExchange,
+        address zeroExProxy,
+        address zrxToken,
+        address[] trustedMsgSenders
+    )
+        public
+    {
+        ZERO_EX_EXCHANGE = zeroExExchange;
+        ZERO_EX_TOKEN_PROXY = zeroExProxy;
+        ZRX = zrxToken;
+
+        for (uint i = 0; i < trustedMsgSenders.length; i++) {
+            TRUSTED_MSG_SENDER[trustedMsgSenders[i]] = true;
+        }
+
+        // The ZRX token does not decrement allowance if set to MAX_UINT
+        // therefore setting it once to the maximum amount is sufficient
+        // NOTE: this is *not* standard behavior for an ERC20, so do not rely on it for other tokens
+        ZRX.approve(ZERO_EX_TOKEN_PROXY, MathHelpers.maxUint256());
+    }
+
+    // ============ Public Functions ============
+
+    function exchange(
+        address tradeOriginator,
+        address receiver,
+        address makerToken,
+        address takerToken,
+        uint256 requestedFillAmount,
+        bytes orderData
+    )
+        external
+        returns (uint256)
+    {
+        // prepare the exchange
+        Order memory order = parseOrder(orderData, makerToken, takerToken);
+        bytes memory signature = parseSignature(orderData);
+
+        // transfer ZRX fee from trader if applicable
+        transferTakerFee(
+            order,
+            tradeOriginator,
+            requestedFillAmount
+        );
+
+        // make sure that the exchange can take the tokens from this contract
+        ensureAllowance(
+            takerToken,
+            ZERO_EX_TOKEN_PROXY,
+            requestedFillAmount
+        );
+
+        // do the exchange
+        IExchange v2Exchange = IExchange(ZERO_EX_EXCHANGE);
+        FillResults memory fill = v2Exchange.fillOrKillOrder(order, requestedFillAmount, signature);
+
+        // validate results
+        assert(fill.takerAssetFilledAmount == requestedFillAmount);
```

Comment:
is it not possible that 0x filled less than the entire `requestedFillAmount`?

Reply:
via `fillOrKill` 0xV2 requires this already

https://github.com/0xProject/0x-monorepo/blob/v2-prototype/packages/contracts/src/2.0.0/protocol/Exchange/MixinWrapperFunctions.sol#L53